File: src/gui/dialogs/SettingsDialog.cpp
//
// Created by Dylan Meadows on 2020-03-15.
//

#include "SettingsDialog.h"

#include <QPushButton>
#include <QTabWidget>
#include <QVBoxLayout>

namespace gui::dialog {
    SettingsDialog::SettingsDialog(
        model::settings::TimerSettingsModel *timerSettings,
        model::settings::ActionSettingsModel *actionSettings, QWidget *parent)
        : QDialog(parent),
          timerSettings(timerSettings),
          actionSettings(actionSettings) {
        initComponents();
    }

    void SettingsDialog::initComponents() {
        setWindowTitle("Preferences");
        setWindowFlags(Qt::Dialog | Qt::WindowTitleHint |
                       Qt::CustomizeWindowHint | Qt::WindowCloseButtonHint);
        auto *layout = new QGridLayout(this);
        layout->setVerticalSpacing(10);
        // ----- tabPane -----
        {
            auto *tabPane = new QTabWidget();
            actionSettingsPane =
                new settings::ActionSettingsPane(actionSettings);
            timerSettingsPane = new settings::TimerSettingsPane(timerSettings);
            tabPane->addTab(actionSettingsPane, "Action");
            tabPane->addTab(timerSettingsPane, "Timer");
            layout->addWidget(tabPane, 0, 0, 1, 2);
        }
        // ----- cancelButton -----
        {
            auto *cancelButton = new QPushButton("Cancel");
            connect(cancelButton, &QPushButton::clicked,
                    [this] { done(QDialog::Rejected); });
            layout->addWidget(cancelButton, 1, 0);
        }
        // ----- okButton -----
        {
            auto *okButton = new QPushButton("OK");
            connect(okButton, &QPushButton::clicked, [this] {
                actionSettingsPane->updateSettings();
                timerSettingsPane->updateSettings();
                done(QDialog::Accepted);
            });
            layout->addWidget(okButton, 1, 1);
            okButton->setDefault(true);
        }
    }
}  // namespace gui::dialog

File: eon_timer/gui/settings_dialog.py
`import functools
import functools
from typing import Optional

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (QDialog, QGridLayout, QPushButton, QTabWidget,
                               QWidget)


class SettingsDialog(QDialog):
    def __init__(self, parent: Optional[QWidget] = ...) -> None:
        super().__init__(parent)
        self.__init_components()

    def __init_components(self) -> None:
        self.setWindowTitle('Preferences')
        self.setWindowFlags(functools.reduce(
            lambda f1, f2: f1 | f2,
            [
                Qt.Dialog,
                Qt.WindowTitleHint,
                Qt.CustomizeWindowHint,
                Qt.WindowCloseButtonHint
            ]))
        layout = QGridLayout(self)
        layout.setVerticalSpacing(10)

        tab_pane = QTabWidget()
        tab_pane.addTab(None, "Action")
        tab_pane.addTab(None, "Timer")
        layout.addWidget(tab_pane, 0, 0, 1, 2)

        ok_btn = QPushButton('OK')
        ok_btn.clicked.connect()

        cancel_btn = QPushButton('Cancel')
        cancel_btn.clicked.connect(lambda _: self.done(QDialog.Rejected))
